common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -0,0 +1,893 @@
+    if (!mergedShuffleFile.exists()) {
+      throw new RuntimeException(String.format("Merged shuffle file %s of %s not found",
+        mergedShuffleFile.getPath(), id.toString()));
+    }
+    File indexFile = getMergedShuffleIndexFile(id);
+    try {
+      // If we get here, the merged shuffle file should have been properly finalized. Thus we can
+      // use the file length to determine the size of the merged shuffle block.
+      ShuffleIndexInformation shuffleIndexInformation = indexCache.get(indexFile);
+      ShuffleIndexRecord shuffleIndexRecord = shuffleIndexInformation.getIndex(chunkId);
+      return new FileSegmentManagedBuffer(
+        conf, mergedShuffleFile, shuffleIndexRecord.getOffset(), shuffleIndexRecord.getLength());
+    } catch (ExecutionException e) {
+      throw new RuntimeException("Failed to open file: " + indexFile, e);
+    }
+  }
+
+  /**
+   * The logic here is consistent with
+   * org.apache.spark.storage.DiskBlockManager#getMergedShuffleFile
+   */
+  // TODO should we use subDirsPerLocalDir to potentially reduce inode size?
+  private File getFile(String appId, String filename) {
+    int hash = JavaUtils.nonNegativeHash(filename);
+    // TODO: Change the message when this service is able to handle NM restart
+    AppPathsInfo appPathsInfo = Preconditions.checkNotNull(appsPathInfo.get(appId),
+      "application " + appId + " is not registered or NM was restarted.");
+    Path[] activeLocalDirs = getActiveLocalDirs(appPathsInfo.activeLocalDirs);
+    Path localDir = activeLocalDirs[hash % activeLocalDirs.length];
+    String relativePath = getRelativePath(appPathsInfo.user, appId);
+    Path filePath = localDir.resolve(relativePath);
+    File targetFile = new File(filePath.toFile(), filename);
+    logger.debug("Get merged file {}", targetFile.getAbsolutePath());
+    return targetFile;
+  }
+
+  private Path[] getActiveLocalDirs(String[] activeLocalDirs) {
+    Preconditions.checkNotNull(activeLocalDirs,
+      "Active local dirs list has not been updated by any executor registration");
+    return Arrays.stream(activeLocalDirs).map(localDir -> Paths.get(localDir)).toArray(Path[]::new);
+  }
+
+  private String getRelativePath(String user, String appId) {
+    return String.format(relativeMergeDirPathPattern + MERGE_MANAGER_DIR, user, appId);
+  }
+
+  private File getMergedShuffleDataFile(AppShufflePartitionId id) {
+    String fileName = String.format("%s.data", id.generateFileName());
+    return getFile(id.appId, fileName);
+  }
+
+  private File getMergedShuffleIndexFile(AppShufflePartitionId id) {
+    String indexName = String.format("%s.index", id.generateFileName());
+    return getFile(id.appId, indexName);
+  }
+
+  private File getMergedShuffleMetaFile(AppShufflePartitionId id) {
+    String metaName = String.format("%s.meta", id.generateFileName());
+    return getFile(id.appId, metaName);
+  }
+
+  @Override
+  public String[] getMergedBlockDirs(String appId) {
```

Comment:
If it's used for test only, please add a comment to it, e.g., "used for test only"

Reply:
It's not used for test only. This will be used when merged shuffle data is fetched which will be part of future PRs.

Reply:
The merged shuffle partition is different from regular shuffle blocks in the sense that its ownership is at the host level, not at the executor level.
There is actually some similarity with SPARK-27651.
When a reducer is fetching merged shuffle partition chunks from its local host, it's similar to fetching node local shuffle blocks in the sense that the reducer needs to know from the local shuffle service about the local dirs for the blocks.
We thus leverage SPARK-27651 to reuse existing code path to handle this part.